src/request.ts

```diff
@@ -0,0 +1,122 @@
+import NetError from './error';
+import {addEventListeners} from './response';
+import {Methods, ParsedHeader, QueryParams, Request, RequestHeaders, RequestOptions} from './types';
+
+export function createRequest(method: Methods) {
+  return function runRequest<T>(
+    url: string,
+    options?: RequestOptions<T>
+  ): Request<T> {
+    return request(method, url, options);
+  };
+}
+
+export function request<T>(
+  method: string,
+  url: string,
+  options: RequestOptions<T> = {}
+): Request<T> {
+  const request = new XMLHttpRequest();
+  request.open(method, parseUrl(url, options.query));
+
+  const headers = parseHeaders(options.headers);
+  headers.forEach(header =>
+    request.setRequestHeader(header.name, header.value)
+  );
+
+  request.withCredentials = true;
+  request.responseType = options.responseType || 'json';
+
+  const cancellable = addEventListeners<T>(request, options);
+
+  const body: any = (options.transformers && options.transformers.response)
+    ? options.transformers.response(options.body)
+    : options.body;
+  request.send(parseRequestBody(body, headers));
+
+  return {
+    request,
+    on: {
+      complete: cancellable._promise
+    },
+    cancel: cancellable.cancel
+  };
+}
+
+function parseUrl(url: string, query?: QueryParams): string {
+  if (query) {
+    const queryString = parseQuery(query);
+    return url.includes('?')
+      ? url.endsWith('&')
+        ? `${url}${queryString}`
+        : `${url}&${queryString}`
+      : `${url}?${queryString}`;
+  } else {
+    return url;
+  }
+}
+
+function parseHeaders(headers?: RequestHeaders): ParsedHeader[] {
+  if (!headers) {
+    return [];
+  } else if (Array.isArray(headers)) {
+    return headers.map(([name, value]) => ({
+      name: normalizeHeaderName(name),
+      value: `${value}`
+    }));
+  } else {
+    return Object.keys(headers).map(name => ({
+      name: normalizeHeaderName(name),
+      value: `${headers[name]}`
+    }));
+  }
+}
+
+function parseRequestBody(body: any, headers: ParsedHeader[]): any | null {
+  if (body) {
+    const contentType = headers
+      .filter(header => header.name == 'Content-Type')
```

Comment:
Is it ok to be using double equals throughout this project?

Reply:
Eh, I'm still on the mindset that you shouldn't compare across types so `===` vs `==` is meaningless 😄 
But, this means we've missed a linting rule to force `===`.  I can add that in.

Reply:
Huh, looks like this is not in the `eslint:recommended` rules like I thought... I would recommend leaving it as is until we create our own fork of create-react-app library setup and add it in there so that we get consistency across code bases.